test/integration/tcpingress_test.go

```diff
@@ -147,3 +150,333 @@ func TestTCPIngressEssentials(t *testing.T) {
+	require.NoError(t, err)
+	defer func() {
+		t.Logf("ensuring that TCPIngress %s is cleaned up", tcpX.Name)
+		if err := c.ConfigurationV1beta1().TCPIngresses(ns.Name).Delete(ctx, tcpX.Name, metav1.DeleteOptions{}); err != nil {
+			if !errors.IsNotFound(err) {
+				require.NoError(t, err)
+			}
+		}
+	}()
+
+	tcpY := &kongv1beta1.TCPIngress{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf(testName, "y"),
+			Namespace: ns.Name,
+			Annotations: map[string]string{
+				annotations.IngressClassKey: ingressClass,
+			},
+		},
+		Spec: kongv1beta1.TCPIngressSpec{
+			Rules: []kongv1beta1.IngressRule{
+				{
+					Host: testServiceSuffixes[2] + ".example",
+					Port: 8899,
+					Backend: kongv1beta1.IngressBackend{
+						ServiceName: testServices[testServiceSuffixes[2]].Name,
+						ServicePort: 1025,
+					},
+				},
+			},
+		},
+	}
+	tcpY, err = c.ConfigurationV1beta1().TCPIngresses(ns.Name).Create(ctx, tcpY, metav1.CreateOptions{})
+	require.NoError(t, err)
+	defer func() {
+		t.Logf("ensuring that TCPIngress %s is cleaned up", tcpY.Name)
+		if err := c.ConfigurationV1beta1().TCPIngresses(ns.Name).Delete(ctx, tcpY.Name, metav1.DeleteOptions{}); err != nil {
+			if !errors.IsNotFound(err) {
+				require.NoError(t, err)
+			}
+		}
+	}()
+
+	for _, i := range testServiceSuffixes {
+		t.Logf("verifying TCP Ingress for %s.example operational", i)
+		require.Eventually(t, func() bool {
+			conn, err := tls.Dial("tcp", fmt.Sprintf("%s:8899", proxyURL.Hostname()), &tls.Config{
+				InsecureSkipVerify: true,
+				ServerName:         fmt.Sprintf("%s.example", i),
+			})
+			if err != nil {
+				return false
+			}
+			defer conn.Close()
+			resp := make([]byte, 512)
+			conn.SetDeadline(time.Now().Add(time.Second * 5))
+			_, err = conn.Read(resp)
+			if err != nil {
+				return false
+			}
+			if strings.Contains(string(resp), i) {
+				return true
+			}
+			return false
+		}, ingressWait, waitTick)
+	}
+
+	tcpY.Spec.Rules[0].Backend.ServiceName = testServiceSuffixes[0]
+	// Update wipes out tcpY if actually assigned, breaking the deferred delete. we have no use for it, so discard it
+	_, err = c.ConfigurationV1beta1().TCPIngresses(ns.Name).Update(ctx, tcpY, metav1.UpdateOptions{})
+	t.Logf("verifying TCP Ingress routes to new upstream after update")
+	require.Eventually(t, func() bool {
+		conn, err := tls.Dial("tcp", fmt.Sprintf("%s:8899", proxyURL.Hostname()), &tls.Config{
+			InsecureSkipVerify: true,
+			ServerName:         fmt.Sprintf("%s.example", testServiceSuffixes[0]),
+		})
+		if err != nil {
+			return false
+		}
+		defer conn.Close()
+		resp := make([]byte, 512)
+		conn.SetDeadline(time.Now().Add(time.Second * 5))
+		_, err = conn.Read(resp)
+		if err != nil {
+			return false
+		}
+		if strings.Contains(string(resp), testServiceSuffixes[0]) {
+			return true
+		}
+		return false
+	}, ingressWait, waitTick)
+}
+
+func TestTCPIngressTLSPassthrough(t *testing.T) {
+	version, err := getKongVersion()
```

Comment:
On one hand, I agree it's important to allow testing KIC with Kong builds that don't report a semver that parses in a way that makes sense.
On the other hand, the failure mode if version getting fails (the test just being skipped silently) is problematic because, if an issue with the version getting machinery ever fails, we can end up with a test that gets never run, without us even noticing.

Here's a suggestion that makes it possible to run tests against builds with broken version reporting and fails visibly if something goes wrong:
* in `getKongVersion`, conditionally override the returned version if an env variable such as `TEST_KONG_VERSION_OVERRIDE` is set to e.g. `2.7.0`
* always fail the test if `getKongVersion` returns anything but a valid semver (possibly being an override)

Such a setting would not only make suppressing the version check opt-in, but also make it possible to control the functional scope we expect to test with the build of Kong under test.

Reply:
c88133d9 to add the override and flip the logic. Rather than failing immediately, this just runs the test anyway if the version is unknown. I figure:
- Unknown versions are often nightlies with the current nightly versioning scheme. They should support new features.
- If the test does fail, you'll see what it couldn't do rather than a generic "dunno whether this is viable to run" failure. If that's expected, you probably should override the version to something below the threshold.

Tests with a max version we may want to handle differently, but we don't have any of those yet.

Reply:
Sounds convincing. LGTM